File: Problems/Algorithms/1268. Search Suggestions System/search_suggestions_system.py
#!/usr/bin/env python3


class TrieNode:
    def __init__(self):
        self.children = {}
        self.suggestions = []


class Trie:
    def __init__(self):
        self.trie = TrieNode()

    def insert(self, word: str):
        node = self.trie
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]

            node.suggestions.append(word)
            node.suggestions.sort()
            while len(node.suggestions) > 3:
                node.suggestions.pop()

    def search(self, word: str) -> List[str]:
        node = self.trie

        ans = []
        for char in word:
            if char not in node.children:
                break
            node = node.children[char]
            ans.append(node.suggestions)

        for _ in range(len(word)-len(ans)):
            ans.append([])
        return ans


class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        trie = Trie()
        for product in products:
            trie.insert(product)
        return trie.search(searchWord)

File: Problems/Algorithms/1268. Search Suggestions System/SearchSuggestionsSystem.kt
class Solution {
    fun suggestedProducts(products: Array<String>, searchWord: String): List<List<String>> {
        products.sort()
        
        val n = products.size
        var start = 0
        val results: MutableList<MutableList<String>> = mutableListOf()
        
        var prefix = StringBuilder()
        for (ch in searchWord) {
            prefix.append(ch)
            
            val curr = lowerBound(products, start, prefix.toString())
            val m = prefix.length
            
            results.add(mutableListOf())
            for (i in curr..minOf(curr+3, n)-1) {
                if (products[i].length < m || products[i].substring(0, m) != prefix.toString()) {
                    break
                }
                
                results.get(results.size-1).add(products[i])
            }
            
            start = curr
        }
        
        return results
    }
    
    private fun lowerBound(products: Array<String>, start: Int, word: String): Int {
        var left = start
        var right = products.size
        
        while (left < right) {
            val mid = left + (right - left) / 2
            
            if (products[mid] >= word) {
                right = mid
            } else {
                left = mid + 1
            }
        }
        
        return left
    }
}

